### PR TITLE
Codebase refactoring, described in #92

### PR DESCRIPTION
**Pull request by zxbodya**
more detailed description in #92 
- es6
- commonjs
- eslint
- split to smaller modules
- some refactoring
- babel build to support older nodejs versions

### CODE REVIEW

**Comment by dylans** [comment]
Thanks @zxbodya , we will review asap!

**Comment by dylans** [comment]
Just a quick note @zxbodya that we haven't forgotten about it. We've started to look at the PR, and should have some feedback shortly.

File: .editorconfig [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# EditorConfig helps developers define and maintain consistent
+# coding styles between different editors and IDEs
+# editorconfig.org
+
+root = true
+
+[*]
+
+indent_style = space
```
Comment (dylans) [inline .editorconfig]: We prefer tabs over spaces (allows an end user to configure the width of indentation that they prefer to see). Other than with package.json where we must conform to npm.

File: .eslintrc [hunk 2]
```diff
@@ -0,0 +1,26 @@
+{
+  "extends": "airbnb-base",
```
Comment (dylans) [inline .eslintrc]: I'm not familiar with what airbnb defines as their base for linting. Could you elaborate please.

File: .eslintrc [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  "env": {
+    "es6": true,
+    "browser": true,
+    "jasmine": true,
```
Comment (dylans) [inline .eslintrc]: We would prefer to use Intern. Intern 4 which will be released this year has been refactored using TS which will make testing ES code without transpilation easy enough to achieve.

File: .eslintrc [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    "node": true
+  },
+  "rules": {
+    "import/prefer-default-export": 0,
+    "import/no-extraneous-dependencies": 0,
+    "import/no-dynamic-require": 1,
```
Comment (dylans) [inline .eslintrc]: I assume this limits non-static analysis? Probably fine, just making sure.

File: package.json [hunk 5]
```diff
@@ -1,42 +1,61 @@
...
-		"through2": "2.0.1"
-	},
-	"devDependencies": {
-		"codecov.io": "0.1.6",
-		"grunt": "^1.0.1",
-		"gulp": "3.9.1",
-		"intern": "^3.3.0"
-	}
+  "name": "remap-istanbul",
+  "version": "0.6.5-pre",
```
Comment (dylans) [inline package.json]: This should probably be 0.8.0-pre

File: package.json [hunk 6]
```diff
@@ -1,42 +1,61 @@
...
+  "description": "A tool for remapping Istanbul coverage via Source Maps",
+  "homepage": "http://dojotoolkit.org",
```
Comment (dylans) [inline package.json]: Copypasta from before... should just point to the GH repo page?

File: package.json [hunk 7]
```diff
@@ -1,42 +1,61 @@
...
+    "babel-core": "^6.17.0",
+    "babel-plugin-transform-runtime": "^6.15.0",
+    "babel-preset-es2015": "^6.16.0",
+    "codecov.io": "0.1.6",
+    "eslint": "^3.7.0",
+    "eslint-config-airbnb-base": "^8.0.0",
+    "eslint-plugin-import": "^1.16.0",
+    "grunt": "^1.0.1",
+    "gulp": "3.9.1",
+    "intern": "^3.3.0",
```
Comment (dylans) [inline package.json]: For now 3.3.2

File: package.json [hunk 8]
```diff
@@ -1,42 +1,61 @@
...
+    "through2": "2.0.1"
+  },
+  "devDependencies": {
+    "babel-cli": "^6.16.0",
+    "babel-core": "^6.17.0",
+    "babel-plugin-transform-runtime": "^6.15.0",
+    "babel-preset-es2015": "^6.16.0",
+    "codecov.io": "0.1.6",
+    "eslint": "^3.7.0",
+    "eslint-config-airbnb-base": "^8.0.0",
```
Comment (dylans) [inline package.json]: Same question as earlier.

File: package.json [hunk 9]
```diff
@@ -1,42 +1,61 @@
...
+    "eslint-plugin-import": "^1.16.0",
+    "grunt": "^1.0.1",
+    "gulp": "3.9.1",
+    "intern": "^3.3.0",
+    "rimraf": "^2.5.4"
```
Comment (dylans) [inline package.json]: New dependency, seems fine, but please comment on what it's used for.

File: src/CoverageTransformer.js [hunk 10]
```diff
@@ -0,0 +1,246 @@
+const { Collector } = require('istanbul');
```
Comment (dylans) [inline src/CoverageTransformer.js]: If we're switching to ES6, we should probably switch to ESM rather than CJS? (I realize that may not be practical for now. Will need feedback from others (@kitsonk or @bryanforbes) if we're ok with this change.

File: src/CoverageTransformer.js [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+const SparceCoverageCollector = require('./SparceCoverageCollector').SparceCoverageCollector;
+
+const getMapping = require('./getMapping');
+
+const remapFunction = require('./remapFunction');
+const remapBranch = require('./remapBranch');
+
+
+class CoverageTransformer {
+  constructor(options) {
```
Comment (dylans) [inline src/CoverageTransformer.js]: earlier comment about tabs vs. spaces... won't repeat myself further on this. :)

File: src/SparceCoverageCollector.js [hunk 12]
```diff
@@ -0,0 +1,119 @@
+class SparceCoverageCollector {
```
Comment (dylans) [inline src/SparceCoverageCollector.js]: We're not generally big fans of ES6 classes because they don't add much and don't provide the safeguards you would expect from classes. That said, that's probably fine here.

File: src/remap.js [hunk 13]
```diff
@@ -0,0 +1,37 @@
+const { CoverageTransformer } = require('./CoverageTransformer');
+
+/**
+ * Remaps coverage data based on the source maps it discovers in the
+ * covered files and returns a coverage Collector that contains the remappped
+ * data.
+ * @param  {Array|Object} coverage The coverage (or array of coverages) that need to be
+ *                                                 remapped
```
Comment (dylans) [inline src/remap.js]: Indentation style here and below seems a bit extreme

File: src/remapBranch.js [hunk 14]
```diff
@@ -0,0 +1,28 @@
+function remapBranch(genItem, getMapping) {
+  const locations = [];
+  let source;
+
+  for (let i = 0; i < genItem.locations.length; i += 1) {
```
Comment (dylans) [inline src/remapBranch.js]: Seems like a good candidate to switch to a for of loop?

File: src/writeReport.js [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+  json: 'file',
+  lcovonly: 'file',
+  teamcity: 'file',
+  'text-lcov': 'console',
+  'text-summary': 'file',
+  text: 'file',
+};
+
+/**
+ * Generates an Instanbul Coverage report based on the information passed.
```
Comment (dylans) [inline src/writeReport.js]: Not a fan of this indentation style. :)

File: typings/browser.d.ts [hunk 16]
```diff
@@ -1 +0,0 @@
-/// <reference path="browser/ambient/node/index.d.ts" />
```
Comment (dylans) [inline typings/browser.d.ts]: Typings files appear to have been deleted. This will need to be regenerated so that TypeScript users may continue to use remap-istanbul.

File: .editorconfig (hunk 1 above)
Comment (zxbodya) [inline .editorconfig]: I prefer spaces :) But I am ok with that - updated code there, changed eslint rules for tabs

File: .eslintrc (hunk 3 above)
Comment (zxbodya) [inline .eslintrc]: that is about eslint checks to allow jasmine things like `describe`… will remove it (it remained from my fork, while rebasing it)

tests itself were mostly not affected in this pull requests

File: .eslintrc (hunk 4 above)
Comment (zxbodya) [inline .eslintrc]: yes, by default it is configured to error lever in airbnb preset this line says - to show just a warning to note that this should be eventually changed(currently we have such require in writeReport.js, and I think we will change that when adding new cli interface)

File: package.json (hunk 5 above)
Comment (zxbodya) [inline package.json]: Ok. Updated it

File: package.json (hunk 6 above)
Comment (zxbodya) [inline package.json]: Yes, I think it is good idea - changed that line

File: package.json (hunk 7 above)
Comment (zxbodya) [inline package.json]: It would be installed - `^3.3.0` effectively is the same as `3.3.*` :)

File: package.json (hunk 9 above)
Comment (zxbodya) [inline package.json]: so what was added:

1. babel and plugins for it, to target older nodejs versions
2. eslint and presets for it to check codestyle
3. rimraf - to clean dist folder before building new version (to prevent accidentally pushing obsolete files to npm)

File: src/CoverageTransformer.js (hunk 10 above)
Comment (zxbodya) [inline src/CoverageTransformer.js]: I would like to use ES6 modules but, from the practical point, at the moment, commonjs are more suitable for this particular case because we can run tests on current nodejs version without traspiling source code… So at least for now I would prefer staying with commonjs

File: src/SparceCoverageCollector.js (hunk 12 above)
Comment (zxbodya) [inline src/SparceCoverageCollector.js]: I consider them as quite handy syntactic sugar and prefer using them over es5 approach or doing some own helpers…

And I like using them for cases where more object oriented approach looks suitable(for example this one)

File: src/remap.js (hunk 13 above)
Comment (zxbodya) [inline src/remap.js]: moved it a bit :)

File: src/remapBranch.js (hunk 14 above)
Comment (zxbodya) [inline src/remapBranch.js]: prefer not using `for of` when not dealing with iterators… typical `for` likely is more performant, and I think in this particular place it can be noticeable

File: typings/browser.d.ts (hunk 16 above)
Comment (zxbodya) [inline typings/browser.d.ts]: Personally I am not typescript user... so I am not quite sure how to do it right way... 
Can you generate those typings?

**Comment by zxbodya** [comment]
> I have somewhat glossed over some of the meat of the code, but that mostly looks unchanged.

That was intentional :)

I think that changes affecting logic and library interfaces should be discussed separately from code style and refactoring.

And planning to suggest them, after having this change set merged.

> Any apologies that this took so long for this initial feedback...

No worry, I perfectly understand that you may not have much time for this.

**Comment by dylans** [comment]
Thanks @zxbodya for addressing all of the feedback... I'll take another look again tomorrow or Friday, but regardless I think this is probably ready to land unless I see something else with a second look.

**Comment by jdonaghue** [comment]
@zxbodya If you have time, in order for us to keep this manageable, could you create a separate PR for your structural changes (conversion to classes)?

We like all of the work that you have done here, but before merging we want to better assess the transition to  classes in this case and make sure that it is something which is in alignment with the SitePen family of open source projects. Most of the changes you have here we want to land right away though and not have held up by that assessment.

So in summary if you could do the following:
* revert the structural changes you made (like `CoverageTransformer.js`) and put them in a separate PR. It is fine if you want to still convert the modules to ES6 (and would be much appreciate) but it would be nice to consider your structural changes in a separate PR. The changes you made to move `lib` to `src` and only create `lib` on build are fine to leave as well.
* put `gruntRemapIstanbul` back into the `tasks` directory. This is needed so that `grunt.loadNpmTasks` can properly load this npm module as a grunt task
* add back the `typings`. You can just put back in place what was previously there
* change the default tab size to 4 in `.editorconfig`
* change the indentation of your code changes (except for within `package.json`) to tabs 

Thanks again for all of your work on this PR!

**Comment by zxbodya** [comment]
Hi @jdonaghue

> If you have time, in order for us to keep this manageable, could you create a separate PR for your structural changes (conversion to classes)?

The main point of this pull-request is to do such changes and get smaller, easier to maintain modules in result.

In order to keep this manageable - all changes there, are not affecting any logic or library public interfaces. Also if you want to review what and how I was changing - just see individual commits in pull requests.

If, you asking about time: I do not have much time to spend on this project - I am doing this mostly, because I have forked this project and already made all this changes(and more) there, and I would like to merge it back, to have one better project in result… But about project itself: it is not likely that I would be using it any time soon - beause of istanbuljs/nyc and coverage support in testing framework I am using now(jest).

---

Completely do not get your point about classes - what is wrong about using them?

You do not like class syntax? But it is already standard part of JavaScript, and I see no reason why avoid using it…

Or question is why to use object oriented approach doing this? - We have some mutable state, methods on it, and some configuration and dependencies… so, using classes looks most oblivious choice to me

----

> revert the structural changes you made (like CoverageTransformer.js) and put them in a separate PR. 

described above: do not agree with this point, at least for now

> It is fine if you want to still convert the modules to ES6 (and would be much appreciate) but it would be nice to consider your structural changes in a separate PR.

explained above, why commonjs…

it is just more practical… at least for now (or we need to do something with tests - to support loading es6 modules... - but is is not intent of this pull-request, and I think - should be done separately)

> put gruntRemapIstanbul back into the tasks directory. This is needed so that grunt.loadNpmTasks can properly load this npm module as a grunt task

there is script in tasks folder for that purpose, task sources were moved to src in order to be compiled for older nodejs versions

> add back the typings. You can just put back in place what was previously there

reverted them

> change the default tab size to 4 in .editorconfig

ok…

> change the indentation of your code changes (except for within package.json) to tabs

that was already done, in previous commit

**Comment by dylans** [comment]
> Completely do not get your point about classes - what is wrong about using them?
> You do not like class syntax? But it is already standard part of JavaScript, and I see no reason why avoid using it…
> Or question is why to use object oriented approach doing this? - We have some mutable state, methods on it, and some configuration and dependencies… so, using classes looks most oblivious choice to me

I think that @jdonaghue was just merely wanting to consider them separately. To be fair, ES6 classes are a bit of a fail, because:

* Support for mixins and traits were dropped
* They're not really OO, they're just sugar around prototype-based inheritance (which is mostly a good thing actually)
* ES6 classes do not yet support properties in the Class prototype
* Ancestors are still mutable with ES6 classes (Anyone can modify the prototype of an ancestor and impact the descent)

That said, they're probably fine for the basic usage here. Mostly we feel that the average engineer expects more from an ES6 Class than it actually provides, especially with regards to mutable ancestors and properties on the Class prototype.

**Comment by zxbodya** [comment]
> Support for mixins and traits were dropped

It is a bit controversial thing… 

I am not fan of mixins, and generally avoid using them…

Instead, I prefer composition of separate instances - it happens to be simpler (having smaller separate things is always easier to manage, also when writing mixin you do not know for sure where and with what it would be mixed… and it can result in conflicts and quite nasty bugs )…

But anyway - there is no barrier for implementing mixins with classes, as with it functions before…

> They're not really OO, they're just sugar around prototype-based inheritance (which is mostly a good thing actually)

Yes, it is syntax sugar, over existing prototypical inheritance… But it is good thing, at least for few reasons:
 - syntax is simpler
 - easier and less error-prone to set up inheritance comparing to using `function` and prototype property
 - constructor will throw an error when called without new keyword

It is not "true" OO in terms of languages like C++ or Java, but it is just different inheritance model, and likely it is good idea not to introduce one more inheritance model in javascript

> ES6 classes do not yet support properties in the Class prototype

Not quite, get what you mean… 
Is it about assigning properties to class prototype, same way as for constructor defined as function? - actually it works the same way as before.

---

In summary: class syntax is already part of the language, they make code cleaner, and are compatible with old code… - so I see no reason why not use them…

**Comment by dylans** [comment]
@zxbodya for what it's worth, I think we're generally in agreement (for example, we support the concept of mixins/traits in https://github.com/dojo/compose , but the way we do it I believe addresses exactly what your objection is to normal usage of mixins).

Anyway, I'm inclined to agree that we just accept this change @jdonaghue since it's not actually changing the way someone will use/consume the API.

**Comment by jdonaghue** [comment]
That's a fair point @dylans. @zxbodya I am going to go ahead and merge this in then. Thanks for your work on this PR!
